zebra-chain/src/orchard/shielded_data.rs

```diff
@@ -0,0 +1,83 @@
+//! Orchard shielded data for `V5` `Transaction`s.
+
+use crate::{
+    amount::Amount,
+    orchard::{tree, Action},
+    primitives::{
+        redpallas::{Binding, Signature, SpendAuth},
+        Halo2Proof,
+    },
+    serialization::{AtLeastOne, TrustedPreallocate},
+};
+
+use std::{
+    cmp::{Eq, PartialEq},
+    fmt::Debug,
+};
+
+/// A bundle of [`Action`] descriptions and signature data.
+#[derive(Clone, Debug, PartialEq, Eq, Deserialize, Serialize)]
+pub struct ShieldedData {
+    /// The orchard flags for this transaction.
+    pub flags: Flags,
+    /// The net value of Orchard spends minus outputs.
+    pub value_balance: Amount,
+    /// The shared anchor for all `Spend`s in this transaction.
+    pub shared_anchor: tree::Root,
+    /// The aggregated zk-SNARK proof for all the actions in this transaction.
+    pub proof: Halo2Proof,
+    /// The Orchard Actions.
+    pub actions: AtLeastOne<AuthorizedAction>,
+    /// A signature on the transaction `sighash`.
+    pub binding_sig: Signature<Binding>,
+}
+
+/// An authorized action description.
+///
+/// Every authorized Orchard `Action` must have a corresponding `SpendAuth` signature.
+#[derive(Clone, Debug, PartialEq, Eq, Deserialize, Serialize)]
+pub struct AuthorizedAction {
+    /// The action description of this Action.
+    pub action: Action,
+    /// The spend signature.
+    pub spend_auth_sig: Signature<SpendAuth>,
+}
+
+impl AuthorizedAction {
+    /// Split out the action and the signature for V5 transaction
+    /// serialization.
+    pub fn into_parts(self) -> (Action, Signature<SpendAuth>) {
+        (self.action, self.spend_auth_sig)
+    }
+}
+
+impl TrustedPreallocate for Action {
+    fn max_allocation() -> u64 {
+        // TODO: fix this
+        1_000_000
+    }
+}
+
+impl TrustedPreallocate for Signature<SpendAuth> {
+    fn max_allocation() -> u64 {
+        // TODO: fix this
+        1_000_000
+    }
+}
```

Comment:
Security: find the maximum serialized sizes of these types, and use them to calculate how many we can safely allocate inside the block limit.

Reply:
`Signature<SpendAuth>` uses the same array size as `Action`, so it doesn't need to implement `TrustedPreallocate`.

According to the spec, `Action`s are `5 * 32 + 580 + 80` bytes and `Signature<SpendAuth>` is `64` bytes, so the total size of each `Action` + `Signature<SpendAuth>` is `884` bytes. (Since there must be a `Signature<SpendAuth>` for each `Action`, we use both their sizes to calculate the limit.)

So the maximum number of `Action`s in a block or transaction is `MAX_BLOCK_BYTES / 884`.

Here is some example code for implementing and testing `TrustedPreallocate`

Size constants:
https://github.com/ZcashFoundation/zebra/blob/2ea1d8624d1d72abdf582953f4f1133efe9eea4f/zebra-chain/src/sprout/joinsplit.rs#L102-L114
Trait impl:
https://github.com/ZcashFoundation/zebra/blob/2ea1d8624d1d72abdf582953f4f1133efe9eea4f/zebra-chain/src/sprout/joinsplit.rs#L122-L129
Size test:
https://github.com/ZcashFoundation/zebra/blob/2ea1d8624d1d72abdf582953f4f1133efe9eea4f/zebra-chain/src/sprout/tests/preallocate.rs#L15-L21
Allocation test:
https://github.com/ZcashFoundation/zebra/blob/main/zebra-chain/src/sprout/tests/preallocate.rs#L33-L63

Reply:
Thank you for the detailed information. I implemented this at https://github.com/ZcashFoundation/zebra/pull/2116/commits/be8f8194e6f9be3c2a8021aad45a22e6adefe2a8 however i think we actually do need the 2 implementations of `TrustedPreallocate`.

With what we have, the build will fail if we dont have them both:

```
error[E0277]: the trait bound `signature::Signature<redpallas::SpendAuth>: TrustedPreallocate` is not satisfied
   --> zebra-chain/src/transaction/serialize.rs:334:13
    |
334 |             zcash_deserialize_external_count(actions.len(), &mut reader)?;
    |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `TrustedPreallocate` is not implemented for `signature::Signature<redpallas::SpendAuth>`
    | 
   ::: zebra-chain/src/serialization/zcash_deserialize.rs:79:76
    |
79  | pub fn zcash_deserialize_external_count<R: io::Read, T: ZcashDeserialize + TrustedPreallocate>(
    |                                                                            ------------------ required by this bound in `zcash_deserialize_external_count`
```